pinata/pinata.go

```diff
@@ -0,0 +1,225 @@
+package pinata
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/opensaucerer/bifrost/shared/errors"
+	"github.com/opensaucerer/bifrost/shared/types"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"os"
+	"path/filepath"
+	 "encoding/json"
+)
+
+type MetaData struct {
+    Name   string      `json:"name"`
+}
+
+type SuccessResponse struct {
+	 IpfsHash   string `json:"IpfsHash"`
+	 Timestamp 	string `json:"Timestamp"`
+	 PinSize 		int64 `json:"PinSize"`
+}
+
+/*
+	Upload File to pinata IPFS
+*/
+
+func (p PinataIPFSStorage) UploadFile(path, filename string, options map[string]interface{}) (*types.UploadedFile, error) {
+
+  // validate JWT	
+	success,err := p.TestAuthentication()
+	if err != nil {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("problem authenticating: %v", err),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+	if !success {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("authentication failed"),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("file does not exist: %s", path),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+
+	url := "https://api.pinata.cloud/pinning/pinFileToIPFS"
+	method := "POST"
+
+	payload := &bytes.Buffer{} // buffer writer
+	writer := multipart.NewWriter(payload)
+	file, fileOpenError := os.Open(path)
+
+	if fileOpenError != nil {
+		return nil, &errors.BifrostError{
+			Err:       fileOpenError,
+			ErrorCode: errors.ErrFileOperationFailed,
+		}
+	}
+
+	part1, err := writer.CreateFormFile("file", filepath.Base(path))
+	if err != nil {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("file does not exist: %s", path),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+
+	_, errFile1 := io.Copy(part1, file)
+	if errFile1 != nil {
+		return nil, &errors.BifrostError{
+			Err:       fmt.Errorf("failed to copy multipart data"),
+			ErrorCode: errors.ErrBadRequest,
+		}
+	}
+
+	defer file.Close()
+
+	 metaData := map[string]string{"name": filename}
+	 pinataMetadata,err := json.Marshal(metaData)
+	 if err!= nil {
+		 return nil, &errors.BifrostError{
+        Err:       fmt.Errorf("failed to marshal metadata"),
+        ErrorCode: errors.ErrBadRequest,
+      }
+	 }
+
+	 pinataOptions,err := json.Marshal(options)
+
+
+	 if err!= nil {
+		 return nil, &errors.BifrostError{
+        Err:       fmt.Errorf("failed to convert metadata to JSON"),
+        ErrorCode: errors.ErrBadRequest,
+      }
+	 }
+
+  _ = writer.WriteField("pinataOptions", string(pinataOptions))
+  _ = writer.WriteField("pinataMetadata", string(pinataMetadata))
+
+	err = writer.Close()
+  if err != nil {
+    return nil, &errors.BifrostError{
+      Err:       fmt.Errorf("failed to close multipart writer"),
+      ErrorCode: errors.ErrBadRequest,
+    }
+  }
+
+	client := &http.Client {
+  }
+  req, err := http.NewRequest(method, url, payload)
```

Comment:
Don't forget that users are able to specify a default time out, this means your request should make use of context with a timeout `NewRequestWithContext` but please check the comment added to the `Disconnect()` function for better clarity on how to migrate the code here and make things much more reusable.

Reply:
> Thank you very much for this contribution, chief. Just a few comments. Please don't be discouraged by these comments. The idea is to build something scalable and easy for everyone to use.

Yea man, it's a lot but i'm really learning new stuffs and idiomatic way of wring go programs